proposals/break-continue-in-inline-lambdas.md

```diff
@@ -0,0 +1,175 @@
+* **A**: Because Kotlin doesn't require labels for non-local `return`. For sake of consistency, We should either require labels
+  for both `break`/`continue` and `return` cases or don't require the labels again for the both cases.
+
+* **Q**: Ok, why not make labels mandatory for both non-local `return` and non-local `break`/`continue` then?
+* **A**: We can't make non-local `return` to be obligatory labeled because it would break inline functions first-class
+  citizenship in language constructs. E.g. we value "return if not null" idiomatic Kotlin construct `nullable?.let { return it }`.
+  If non-local `return` were obligatory labeled then the code would be much more verbose
+  `nullable?.let { return@functionName it }`.
+
+It worth noticing that we don't abandon the idea to provide some "break"/"continue" mechanism for "loop-like" functions
+[KT-19748](https://youtrack.jetbrains.com/issue/KT-19748). It just that we agree that we will never support unlabeled non-local
+`break`/`continue` for "loop-like" functions.
+
+## Tooling for puzzling code
+
+Some users may find constructions containing `break` or `continue` within lambdas passed to "loop-like" inline functions
+such as `forEach` or `takeWhile` puzzling:
+```kotlin
+fun printUntilZero(producer: () -> List<Int>) {
+    while(true) {
+        val list = producer()
+        list.forEach {
+            if (it == 0) break // Does it exit from while or from forEach?
+            println(it)
+        }
+    }
+}
+```
+
+To reduce the risk of confusion, it was proposed to introduce an IDE inspection and a quickfix recommending to use labeled `break`
+and `continue` in lambdas passed to parameters without an `EXACTLY_ONCE` or `AT_MOST_ONCE` contract (potentially "loop-like").
+After the inspection is applied:
+```kotlin
+fun printUntilZero(producer: () -> List<Int>) {
+    myLoop@ while(true) {
+        val list = producer()
+        list.forEach {
+            if (it == 0) break@myLoop
+            println(it)
+        }
+    }
+}
+```
+
+The inspection should not triggerred for `let`, `run`, etc as they have the corresponding contract. There seems to be no need for
+a similar inspection for non-local returns. We assume that they are less confusing.
+
+## Implementation note
+
+Similarly to non-local returns, non-local break/continue will only work for lambdas passed as parameters not marked
+with `noinline` and `crossinline`.
+
+Non-local `break` and `continue` could be emulated using non-local return statements.
+For example, the following code snippets should be equivalent with respect to DFA:
+
+```kotlin
+for (i in 0 until 100) {
+    run {
+        if (i == 3) break
+    }
+}
+```
+```kotlin
+run breaking@{
+    for (i in 0 until 100) {
+        run {
+            if (i == 3) return@breaking
+        }
+    }
+}
+```
+
+## Potential issues with data-flow analysis
+
+We should check that the new feature doesn't introduce new data-flow analysis issues:
```

Comment:
I just realized that I don't like "action items" in our KEEP documents. IMHO, KEEP document should be a document that tries to convince that the feature is worth to be added. 

Action items in such kind of document only raise a question: "so what did someone do the action item"

So :) Did we check that non-local `break` and `continue` don't introduce data-flow analysis issues?